packages/jupyter-ai/jupyter_ai/personas/base_persona.py

```diff
@@ -0,0 +1,189 @@
+from abc import ABC, abstractmethod
+from dataclasses import asdict
+from logging import Logger
+from time import time
+from typing import TYPE_CHECKING, Any, Dict, Optional, Set
+
+from jupyter_ai.config_manager import ConfigManager
+from jupyterlab_chat.models import Message, NewMessage, User
+from jupyterlab_chat.ychat import YChat
+from pydantic import BaseModel
+
+from .persona_awareness import PersonaAwareness
+
+# prevents a circular import
+# `PersonaManager` types have to be surrounded in single quotes
+if TYPE_CHECKING:
+    from collections.abc import AsyncIterator
+
+    from .persona_manager import PersonaManager
+
+
+class PersonaDefaults(BaseModel):
+    """
+    Data structure that represents the default settings of a persona. Each persona
+    must define some basic default settings, like its name.
+
+    Each of these settings can be overwritten through the settings UI.
+    """
+
+    ################################################
+    # required fields
+    ################################################
+    name: str  # e.g. "Jupyternaut"
+    description: str  # e.g. "..."
+    avatar_path: str  # e.g. /avatars/jupyternaut.svg
+    system_prompt: str  # e.g. "You are a language model named..."
+
+    ################################################
+    # optional fields
+    ################################################
+    slash_commands: Set[str] = set("*")  # change this to enable/disable slash commands
+    model_uid: Optional[str] = None  # e.g. "ollama:deepseek-coder-v2"
+    # ^^^ set this to automatically default to a model after a fresh start, no config file
+
+
+class BasePersona(ABC):
+    """
+    Abstract base class that defines a persona when implemented.
+    """
+
+    ychat: YChat
+    manager: "PersonaManager"
+    config: ConfigManager
+    log: Logger
+    awareness: PersonaAwareness
+
+    ################################################
+    # constructor
+    ################################################
+    def __init__(
+        self,
+        *,
+        ychat: YChat,
+        manager: "PersonaManager",
+        config: ConfigManager,
+        log: Logger,
+    ):
+        self.ychat = ychat
+        self.manager = manager
+        self.config = config
+        self.log = log
+        self.awareness = PersonaAwareness(
+            ychat=self.ychat, log=self.log, user=self.as_user()
+        )
+
+        self.ychat.set_user(self.as_user())
+
+    ################################################
+    # abstract methods, required by subclasses.
+    ################################################
+    @property
+    @abstractmethod
+    def defaults(self) -> PersonaDefaults:
+        """
+        Must return a PersonaDefaults object that represents the default
+        settings of this persona.
+        """
+        pass
```

Comment:
how about `raise NotImplementedError` if this method must be implemented by child classes?

Reply:
Abstract classes should extend the `abc.ABC` class in Python, part of the stdlib. That automatically enforces abstract method implementation in child classes:

<img width="880" alt="Screenshot 2025-04-16 at 10 26 20 AM" src="https://github.com/user-attachments/assets/025bf238-c9e7-4869-aae8-892f2312accb" />